@xen-orchestra/audit-core/src/index.js

```diff
@@ -0,0 +1,176 @@
+import assert from 'assert'
+import createLogger from '@xen-orchestra/log'
+import hash from 'object-hash'
+import synchronized from 'decorator-synchronized'
+import { invert, omit, sortBy } from 'lodash'
+
+const log = createLogger('xo:audit-core')
+
+// Format: $<algorithm>$<salt>$<encrypted>
+//
+// http://man7.org/linux/man-pages/man3/crypt.3.html#NOTES
+const ALGORITHM_TO_ID = {
+  md5: '1',
+  sha256: '5',
+  sha512: '6',
+}
+
+const ID_TO_ALGORITHM = invert(ALGORITHM_TO_ID)
+
+const synchronizedFns = synchronized()
+
+export class AuditCore {
+  constructor(storage, { algorithm, retention }) {
+    assert(storage !== undefined && algorithm !== undefined)
+
+    this._algorithm = algorithm
+    this._retention = retention
+    this._storage = storage
+  }
+
+  _createHash(data, algorithm = this._algorithm) {
+    return `$${ALGORITHM_TO_ID[algorithm]}$$${hash(data, {
+      algorithm,
+    })}`
+  }
+
+  @synchronizedFns
+  async add(subject, event, data) {
+    const record = {
+      data,
+      event,
+      previousHash: (await this.getData()).hashes.pop(),
+      subject,
+      time: Date.now(),
+    }
+    record.id = this._createHash(record)
+    await this._storage.add(record.id, record)
+    return record
+  }
+
+  async checkIntegrity(startHash, endHash) {
```

Comment:
I think we should better names than start and end because it's unclear in which direction they are.

Example : `oldest`, `newest`.